cmd/webhook/app/options/options.go

```diff
@@ -2,8 +2,11 @@ package options
 
 import (
 	"github.com/spf13/pflag"
+	"k8s.io/component-base/cli/globalflag"
 
 	"github.com/karmada-io/karmada/pkg/sharedcli/profileflag"
+	// ensure libs have a chance to globally register their flags
+	_ "github.com/karmada-io/karmada/pkg/webhook/defaulttolerationseconds"
```

Comment:
Here, we don't use anything defined at `pkg/webhook/defaulttolerationseconds`. right?

Reply:
We want to initialize `defaultNotReadyTolerationSeconds` and `defaultUnreachableTolerationSeconds` first. So before `AddCustomGlobalFlags`, the `flag.CommandLine` now already has these two flags. When we call `AddCustomGlobalFlags`, these two flags are retrieved from `flag.CommandLine` and inject into `pflag.FlagSet`, please refer to `k8s.io/component-base/cli/globalflag`.

Reply:
I know, when importing `"github.com/karmada-io/karmada/pkg/webhook/defaulttolerationseconds"`, the two flags will be added to `commandline`, then `globalflag.Register(flags,...)` moves the two flags to our FlagSet, right?

It's a little bit tricky, isn't it? Why not just add the two flags in this file?

Reply:
Because it's ugly to pass these parameters into the struct `MutatingAdmission`. We use these two flags to initialize variables`notReadyToleration` and `unreachableToleration`. If we pass these the flags, we could not new the variables in advance and we must set this field every time when mutating function is called. I think this is why `k/k` also does.

Reply:
>  If we pass these the flags, we could not new the variables in advance and we must set this field every time when mutating function is called.

I don't understand why you think `set this field every time`, do you mean the [webhook.Admission](https://github.com/karmada-io/karmada/blob/e1536144771e5a67b9f1043feea97a225e9edd88/cmd/webhook/app/webhook.go#L114) will be build each HTTP call?

> Because it's ugly to pass these parameters into the struct MutatingAdmission.

I made a try based on your commit(https://github.com/karmada-io/karmada/commit/9a4821db6ba34844170b27d74a8c78719efa740e), is it the way(pass parameters) you think ugly?

Reply:
> I don't understand why you think set this field every time, do you mean the [webhook.Admission](https://github.com/karmada-io/karmada/blob/e1536144771e5a67b9f1043feea97a225e9edd88/cmd/webhook/app/webhook.go#L114) will be build each HTTP call?

Right, https://github.com/karmada-io/karmada/commit/9a4821db6ba34844170b27d74a8c78719efa740e generate default toleration every time when handle each HTTP request. Acutually, this toleration will never change since we have set the flag.

I guess the flag is more like a kind of global ones. Once it is set, the toleration is determined. So pp and cpp mutating webhook could use these tolerations directly.

Reply:
> I guess the flag is more like a kind of global ones. Once it is set, the toleration is determined. 

Yes, they are initialized only once.

> Right, https://github.com/karmada-io/karmada/commit/9a4821db6ba34844170b27d74a8c78719efa740e generate default toleration every time when handle each HTTP request.

You are right, I just made another patch at https://github.com/karmada-io/karmada/commit/b25b6e74b784379723ff1a6ab02105340866fb74, i guess the handler will be shared by all HTTP call.

By the way, the reason why I don't like initializing flags by `init` is when people import something(like functions) from the package they don't know the flags are added. I think it's a bad programming model, I have some lessons learned before.
For example, the controller-runtime framework will add `--kubeconfig` flag by default(we ever don't know it since we just import some functions from the package), the flag is useful for most cases, but doesn't make sense to us, we have to modify it, like:
https://github.com/karmada-io/karmada/blob/e1536144771e5a67b9f1043feea97a225e9edd88/cmd/controller-manager/app/controllermanager.go#L80-L84

> I think this is why k/k also does.

In addition, you mentioned Kubernetes did the same thing, I think the reason is the flag is defined and used by a specific admission, so they put the flags in the admission for maintainability. Another reason is the `Plugin` is shared by all admissions, so introducing fields into it doesn't make sense to other admissions.

Reply:
Thanks very much, I got it. I agree with patch https://github.com/karmada-io/karmada/commit/b25b6e74b784379723ff1a6ab02105340866fb74. How can I patch the commit to this PR?

Reply:
This patch is a draft one only used for explaining my idea, only for your reference. (Didn't solve golint issues.)

You can fetch it from my forked repo and cherry-pick it if you want, like:
```
$ git remote add hongcai https://github.com/RainbowMango/karmada.git
$ git fetch hongcai
$ git cherry-pick b25b6e74b784379723ff1a6ab02105340866fb74
```